src/miner.cpp

```diff
@@ -645,42 +663,50 @@ void BitcoinMiner(CWallet* pwallet, bool fProofOfStake)
     bool fLastLoopOrphan = false;
     while (fGenerateBitcoins || fProofOfStake) {
         if (fProofOfStake) {
+            if (chainActive.Tip()->nHeight < Params().LAST_POW_BLOCK()) {
+                //  The last PoW block hasn't even been mined yet.
+                MilliSleep(Params().TargetSpacing() * 1000);       // sleep a block
+                continue;
+            }
+
             //control the amount of times the client will check for mintable coins
             if ((GetTime() - nMintableLastCheck > 5 * 60)) // 5 minute check time
             {
                 nMintableLastCheck = GetTime();
                 fMintableCoins = pwallet->MintableCoins();
             }
 
-            if (chainActive.Tip()->nHeight < Params().LAST_POW_BLOCK()) {
-                MilliSleep(5000);
-                continue;
-            }
-
-            while (vNodes.empty() || pwallet->IsLocked() || !fMintableCoins || (pwallet->GetBalance() > 0 && nReserveBalance >= pwallet->GetBalance()) || !masternodeSync.IsSynced()) {
+            while (vNodes.empty() || pwallet->IsLocked() || !fMintableCoins ||
+                   (pwallet->GetBalance() > 0 && nReserveBalance >= pwallet->GetBalance()) || !masternodeSync.IsSynced()) {
                 nLastCoinStakeSearchInterval = 0;
+                MilliSleep(5000);
                 // Do a separate 1 minute check here to ensure fMintableCoins is updated
-                if (!fMintableCoins) {
-                    if (GetTime() - nMintableLastCheck > 1 * 60) // 1 minute check time
-                    {
-                        nMintableLastCheck = GetTime();
-                        fMintableCoins = pwallet->MintableCoins();
-                    }
+                if (!fMintableCoins && (GetTime() - nMintableLastCheck > 1 * 60)) // 1 minute check time
+                {
+                    nMintableLastCheck = GetTime();
+                    fMintableCoins = pwallet->MintableCoins();
                 }
-                MilliSleep(5000);
-                if (!fGenerateBitcoins && !fProofOfStake)
-                    continue;
```

Comment:
So that's replaced by exiting the thread when PoW has ended, right?

Reply:
Not even...  It's useless on so many different levels; but let me just focus on the continue.

```
while (x) {
  do something;
  while (y) {
     do something else
     if (conditional) continue;
     do something more.
  }
}
```

In the above case; the continue is useful.  "If conditional, restart the while loop "y", otherwise keep going and do something more.

However the case in this actual section of code was:
```
while (x) {
  do something;
  while (y) {
     do something else
     if (conditional) continue;
  }
}
```
If conditional; go back to the beginning of the while y loop.  and skip... nothing.  else.. go back to the beginning of the while y loop.

I don't know if the original intent of the code was to actually continue the "while x" loop, but it was actually continuing the "while y" loop... and any compiler worth it's salt would have optimized those two lines away anyway,

Reply:
Right.